diff --git a/include/v8.h b/include/v8.h
index 18d72f1630c..705b1a2ed89 100644
--- a/include/v8.h
+++ b/include/v8.h
@@ -12040,4 +12040,14 @@ size_t SnapshotCreator::AddData(Local<T> object) {
 
 }  // namespace v8
 
+#define V8_USE_UE_TRACE 1
+namespace v8
+{
+// do not new two ArrayBuffer with the same data and length
+typedef void (*V8CustomUETraceFunction)(const char*);
+V8_EXPORT void SetUETraceBeginFunction(V8CustomUETraceFunction Func);
+V8_EXPORT void SetUETraceEndFunction(V8CustomUETraceFunction Func);
+}
+
+
 #endif  // INCLUDE_V8_H_
diff --git a/src/tracing/trace-event.cc b/src/tracing/trace-event.cc
index 8e69fe55205..ac32fe31be0 100644
--- a/src/tracing/trace-event.cc
+++ b/src/tracing/trace-event.cc
@@ -14,7 +14,8 @@
 namespace v8 {
 namespace internal {
 namespace tracing {
-
+V8CustomUETraceFunction _InternalV8CustomUETraceBegin = nullptr;
+V8CustomUETraceFunction _InternalV8CustomUETraceEnd = nullptr;
 #if !defined(V8_USE_PERFETTO)
 v8::TracingController* TraceEventHelper::GetTracingController() {
   return v8::internal::V8::GetCurrentPlatform()->GetTracingController();
@@ -56,4 +57,14 @@ void CallStatsScopedTracer::Initialize(v8::internal::Isolate* isolate,
 
 }  // namespace tracing
 }  // namespace internal
+
+  void SetUETraceBeginFunction(V8CustomUETraceFunction Func)
+  {
+    internal::tracing::_InternalV8CustomUETraceBegin = Func;
+  }
+  void SetUETraceEndFunction(V8CustomUETraceFunction Func)
+  {
+    internal::tracing::_InternalV8CustomUETraceEnd = Func;
+  }
+
 }  // namespace v8
diff --git a/src/tracing/trace-event.h b/src/tracing/trace-event.h
index 816e79c5232..b402061b03c 100644
--- a/src/tracing/trace-event.h
+++ b/src/tracing/trace-event.h
@@ -16,6 +16,7 @@
 #endif  // !defined(V8_USE_PERFETTO)
 
 #include "include/v8-platform.h"
+#include "include/v8.h"
 #include "src/base/atomicops.h"
 #include "src/base/macros.h"
 
@@ -188,6 +189,7 @@ enum CategoryGroupEnabledFlags {
 #define INTERNAL_TRACE_EVENT_ADD_SCOPED(category_group, name, ...)           \
   INTERNAL_TRACE_EVENT_GET_CATEGORY_INFO(category_group);                    \
   v8::internal::tracing::ScopedTracer INTERNAL_TRACE_EVENT_UID(tracer);      \
+  v8::internal::tracing::UEStatsScopedTracer INTERNAL_TRACE_EVENT_UID(uetracer)(name);   \
   if (INTERNAL_TRACE_EVENT_CATEGORY_GROUP_ENABLED_FOR_RECORDING_MODE()) {    \
     uint64_t h = v8::internal::tracing::AddTraceEvent(                       \
         TRACE_EVENT_PHASE_COMPLETE,                                          \
@@ -203,6 +205,7 @@ enum CategoryGroupEnabledFlags {
                                                   bind_id, flow_flags, ...) \
   INTERNAL_TRACE_EVENT_GET_CATEGORY_INFO(category_group);                   \
   v8::internal::tracing::ScopedTracer INTERNAL_TRACE_EVENT_UID(tracer);     \
+  v8::internal::tracing::UEStatsScopedTracer INTERNAL_TRACE_EVENT_UID(uetracer)(name);   \
   if (INTERNAL_TRACE_EVENT_CATEGORY_GROUP_ENABLED_FOR_RECORDING_MODE()) {   \
     unsigned int trace_event_flags = flow_flags;                            \
     v8::internal::tracing::TraceID trace_event_bind_id(bind_id,             \
@@ -279,6 +282,7 @@ enum CategoryGroupEnabledFlags {
   INTERNAL_TRACE_EVENT_GET_CATEGORY_INFO(category_group);                      \
   v8::internal::tracing::CallStatsScopedTracer INTERNAL_TRACE_EVENT_UID(       \
       tracer);                                                                 \
+  v8::internal::tracing::UEStatsScopedTracer INTERNAL_TRACE_EVENT_UID(uetracer)(name);   \
   if (INTERNAL_TRACE_EVENT_CATEGORY_GROUP_ENABLED_FOR_RECORDING_MODE()) {      \
     INTERNAL_TRACE_EVENT_UID(tracer)                                           \
         .Initialize(isolate, INTERNAL_TRACE_EVENT_UID(category_group_enabled), \
@@ -291,7 +295,8 @@ namespace internal {
 class Isolate;
 
 namespace tracing {
-
+extern V8CustomUETraceFunction _InternalV8CustomUETraceBegin;
+extern V8CustomUETraceFunction _InternalV8CustomUETraceEnd;
 // Specify these values when the corresponding argument of AddTraceEvent
 // is not used.
 const int kZeroNumArgs = 0;
@@ -609,6 +614,16 @@ class CallStatsScopedTracer {
   Data data_;
 };
 
+class UEStatsScopedTracer{
+public:
+  UEStatsScopedTracer(const char* name){
+    if(_InternalV8CustomUETraceBegin) _InternalV8CustomUETraceBegin(name);
+  }
+  ~UEStatsScopedTracer(){
+    if(_InternalV8CustomUETraceEnd) _InternalV8CustomUETraceEnd(nullptr);
+  }
+};
+
 }  // namespace tracing
 }  // namespace internal
 }  // namespace v8
