diff --git a/src/flags/flags.cc b/src/flags/flags.cc
index f43715ac32e..6449b2a9780 100644
--- a/src/flags/flags.cc
+++ b/src/flags/flags.cc
@@ -530,7 +530,7 @@ std::ostream& operator<<(std::ostream& os, const Flag& flag) {
 static std::atomic<uint32_t> flag_hash{0};
 static std::atomic<bool> flags_frozen{false};
 
-uint32_t ComputeFlagListHash() {
+__attribute__((no_sanitize_address)) uint32_t ComputeFlagListHash() {
   std::ostringstream modified_args_as_string;
   if (COMPRESS_POINTERS_BOOL) modified_args_as_string << "ptr-compr";
   if (DEBUG_BOOL) modified_args_as_string << "debug";
@@ -545,7 +545,9 @@ uint32_t ComputeFlagListHash() {
     if (flag.PointsTo(&v8_flags.predictable)) continue;
     modified_args_as_string << flag;
   }
+  
   std::string args(modified_args_as_string.str());
+  
   // Generate a hash that is not 0.
   uint32_t hash = static_cast<uint32_t>(base::hash_range(
                       args.c_str(), args.c_str() + args.length())) |
